### ISSUE TITLE
Taskbar becomes confusing after 18+ windows / order of windows is non-obvious

### ISSUE THREAD

**Issue by ddevz** [QubesOS/qubes-issues#6662]
**The problem you're addressing (if any)**
The taskbar works great when you start out and have 10 windows open.  They tend to group in some kind of way that helps one intuitively tend to find the area of the bar they need to be looking at more then a "first opened" would, then one can just read the names to find which is the correct one.  All is good.

However, when many windows are open, you become unable to read the contents of a window without hovering.  It seems to group by color-disposable (I.E. disposable red counts as separate from non-disposable red), as well as by application.  But the precise sort order escapes me and a web search was not fruitful. 

(note: The "UX" label belongs on this, but I did not see how to add labels)

**Describe the solution you'd like**
I don't have anything in mind.  While communicating what the sort order is is definitely possible, the overall issue of being able to find things easily is not necessarily solvable. I'm writing this because it seems like the type of information that your UX people want.

The solution to this may be related to the recent survey, but the survey is closed and I was unable to re-find the survey videos to confirm.

**Where is the value to a user, and who might that user be?**

**Describe alternatives you've considered**

**Additional context**

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

**Comment by ninavizz** [comment]
Hi @ddevz! Thank you for creating this issue. :) 

I actually created #6654 to track a different but related problem. I also saw the problem you created this issue to solve for, and suggested "Task Bar Widgets" as a solution opportunity for this. Would love to get your thoughts!

From ##6654:
The "Creo" VM has 5 windows open within it, which are revealed when its Task Bar Widget is "opened." 
> Approach n°2: The more resource-intensive solution that would require a generous volunteer or money falling from the sky, would be fully custom per-qube widgets in the Task Bar—of which, the below is a quickie mockup.
> ![image](https://user-images.githubusercontent.com/8262612/120127026-fa9da500-c172-11eb-8106-b60fdd76c060.png)

I have not spent much (cough, any) time exploring solution opportunities to best mitigate the problem you created this issue to address, but the above is one solution that could potentially work well for this. I'd love to get your thoughts! 

The "visual details" of execution, I'd absolutely want to do more exploration/polishing work on, as the approach shown in the above mockup feels rough and in need of refinement... but the core functional idea is demonstrated, acceptably.

**Comment by ddevz** [comment]
> I have not spent much (cough, any) time exploring solution opportunities to best mitigate the problem you created this issue to address, but the above is one solution that could potentially work well for this. I'd love to get your thoughts!

If I understand, you are displaying the menu, and indicating which applications are running by making those bold, and then also putting the tasks for just that VM in the taskbar (which allows one to select *which* of the 3 "creo parametric" windows they want to select).

That's neat.  I like it!  I think that would help the problem.  I'd ask for the virtual machines to get sorted in kind of deterministic order that is able to be learned within a semi-reasonable effort, as i can image 18+ virtual machines being open open in the taskbar once the taskbar is easier to navigate, creating the next issue. :)  I.E.  I don't think it's a "solved"/"not solved" issue, I think it's more of a "how many open windows can a average user effectively manage before they get lost?"

If we suppose that our current maximum number of open windows before getting lost is 18, then i feel your system would push the number much higher... by a factor of 2 or 3 (whatever the average number of windows open per virtual machine normally is) making it 36 to 54.

Deterministic, learnable sorting could potentially push it farther then that. (i don't know how much)

On another note, I can also imagine some people wanting the taskbar menus to be by application, then by VM.  I.E. if one does not remember which VMs are running firefox, then looking through the whole list of VM's to try to remember might be undesirable.  (but I think that is partially related to what the recent survey was about, so you've probably already thought about that case.)

It's also possible that there are organizing principals that one can implement in their workflow to help the situation if we wanted to create a "recommended workflow".

Lastly, you mention it'll probably be a while before your taskbar system gets implemented.  In the meantime it might make sense to just publish what the current taskbar sort order algorithm is until then.

**Comment by andrewdavidwong** [comment]
@ddevz: Have you tried playing with the Xfce settings? You should be able to configure some of this already and learn more about how it currently works.

**Comment by ddevz** [comment]
Wow, you must read every line of every comment of every issue, as that was the last line of a long ramble.  Impressive.

I'll take your comment to mean "Qubes does not intentionally do anything to change whatever the default sort order in Xfce is".  If that's the case then I'll look at the Xfce settings and any documentation.  Thanks.

**Comment by andrewdavidwong** [comment]
> I'll take your comment to mean "Qubes does not intentionally do anything to change whatever the default sort order in Xfce is".  If that's the case then I'll look at the Xfce settings and any documentation.  Thanks.
> 

That's probably accurate but isn't exactly what I meant. I just seem to recall seeing various options for controlling the ordering and/grouping of items on the task bar, so at least some of what you want may already be possible right now.

**Comment by brendanhoar** [comment]
With xfce there's also the ability to increase the number of desktops from the default of four to many more. I tend to keep more then at a time.

**Comment by ddevz** [comment]
> That's probably accurate but isn't exactly what I meant. I just seem to recall seeing various options for controlling the ordering and/or grouping of items on the task bar, so at least some of what you want may already be possible right now.

You are correct, "Qubes start icon" -> System Tools -> Panel -> Items, selecting "Window Buttons" then hitting "Edit" reveals this:

Sorting order:  "Group Title and timestamp"

(it's selectable, with that as the default)

I'll research what it counts as a "group".
Thanks

**Comment by ddevz** [comment]
> With xfce there's also the ability to increase the number of desktops from the default of four to many more. I tend to keep more then at a time.

Your comment brings up some interesting points:

1. when I get multiple monitors operational, the taskbar entries might spread out to the appropriate monitor, changing the situation.
2. @ninavizz , have you thought through how your system would handle the case when half of a VM's windows are on desktop 1, and the other half are on desktop 2?  (and then the same question again but with "monitor 1" and "monitor 2" instead of "desktop 1" and "desktop 2")

**Comment by ninavizz** [comment]
> @ninavizz , have you thought through how your system would handle the case when half of a VM's windows are on desktop 1, and the other half are on desktop 2? (and then the same question again but with "monitor 1" and "monitor 2" instead of "desktop 1" and "desktop 2")

I just came up with that idea off the cuff. A proper design exploration would require some more in depth research, and would also produce multiple paths to a centrally defined problem. The only "problem" the sketch I created seeks to solve for, is how to facilitate easier management of a multi-environment system that is both user friendly and influences choices to not over-stress limited system resources. 

"Spaces" (or "Workspaces") was developed for Windows, I believe, and presumes a very different use scenarios and central use problems. It was developed to facilitate contextual grouping of apps by task needs, not security—so trying to bend it to work with Qubes, feels like trying to wedge a banana into an ice cream cone.

I also don't ever use the "Spaces" feature, and feel we'd need to do some research to learn how many people do, vs how many do not, and with Qubes, how/why they use Spaces.

**Comment by ddevz** [comment]
> > @ninavizz , have you thought through how your system would handle the case when half of a VM's windows are on desktop 1, and the other half are on desktop 2? (and then the same question again but with "monitor 1" and "monitor 2" instead of "desktop 1" and "desktop 2")
> 
> I haven't thought of "how the system would handle that, specifically" because I have no knowledge of why a user would desire that, in the first place. It's important to understand why users want certain things, before extending a feature to elegantly address a stress-case. Otherwise it just turns into a paint-by-numbers mess, like Libre Office's customization options.
> 
> I also don't ever use the "Spaces" feature, and feel we'd need to do some research to learn how many people do, vs how many do not, and with Qubes, how/why they use Spaces.

@brendanhoar was proposing that virtual workspaces could be used as a partial solution to the "confusing after 18+ windows" issue, as the windows on each workspace only show up in the taskbar for *that* workspace, thus reducing the visible number at any one time.

After reading his proposal, I started using the virtual workspaces to see how much it could help.  Since it requires a click to get to the next workspace, then a click on the taskbar, it might not be that convenient.

As for multi-monitor, I have not yet determined what video card to get that will support 6 monitors and work well with qubes, but once I do, using virtual workspaces will make no sense anymore, and I'll be able to report back how the xfce system handles the taskbar across multiple monitors.  (I.E does one big taskbar span all the monitors, or does each monitor have it's own taskbar, or is there just one taskbar on one monitor for all the windows on all monitors?)

Note that people using dual monitors nowdays is common. (but using 6 monitors is not).

**Comment by brendanhoar** [comment]
> @brendanhoar was proposing that virtual workspaces could be used as a partial solution to the "confusing after 18+ windows" issue, as the windows on each workspace only show up in the taskbar for *that* workspace, thus reducing the visible number at any one time.
> 

I use ctrl-alt-left-arrow and ctrl-alt-right-arrow to zoom about the workspaces.

> 
> 
> After reading his proposal, I started using the virtual workspaces to see how much it could help.  Since it requires a click to get to the next workspace, then a click on the taskbar, it might not be that convenient.

Yeah I definitely don't use my mouse much for workspaces just keyboard navigation.

B

**Comment by trackroute** [comment]
If you have lots of windows and virtual desktops then you might find "KDE Activities" useful. Activities are a super-set of Virtual Desktops.

**Comment by unman** [comment]
On Mon, Jun 14, 2021 at 06:21:16PM -0700, trackroute wrote:
> If you have lots of windows and virtual desktops then you might find "KDE Activities" useful. Activities are a super-set of Virtual Desktops.
> 

*THIS* - I am constantly promoting Activities on the lists and in the
forum.
You can group windows by Activity, configure Activities to use a
different coloured Background, and use a name to identify the Activity.
Activities cover all monitors, and the task bar shows only windows in
the activity.
You can also use simple scripting to switch activities when opening a
window, and KWin details to ensure that windows appear in the right
Activity.
KDE is very keyboard focussed - you can navigate activities, and move
windows between them, using keyboard shortcuts, as well as all the usual
windows controls.
KDE also has a native menu editor that makes it easy to provide nested
menus, or groupings of shortcuts, for those that like that sort of
thing. Since a huge effort is being focussed on providing something
that is already available in KDE, this is particularly significant.
On the particular point of this issue, the task bar will group windows
by qube.

KDE is a perfect match for Qubes - the reasons why we moved away from
KDE (resources, ugliness), no longer apply n 2021, and I would recommend
it to any Qubes user - it is **far** better than Xfce.

**Comment by ninavizz** [comment]
@unman Ok—you and Demi are officially Team KDE. I'm leaning towards Gnome, but would love to know if you have thoughts about KDE vs Gnome. Also, why do you like KDE so much? I haven't yet had a chance to look at things like localization or accessibility for either, and if you have any insights into either for KDE I'd love to learn!

**Comment by DemiMarie** [comment]
> @unman Ok—you and Demi are officially Team KDE. I'm leaning towards Gnome, but would love to know if you have thoughts about KDE vs Gnome. Also, why do you like KDE so much? I haven't yet had a chance to look at things like localization or accessibility for either, and if you have any insights into either for KDE I'd love to learn!

I prefer KDE purely because implementing Wayland on KDE will likely be much, much simpler than GNOME.  GNOME is *very* opinionated, which can be a good thing, but it also means that customizing GNOME for our needs appears to be much more difficult.  For instance, Mutter’s lack of server-side decoration support on Wayland means that we will need to implement window decorations ourselves, as libdecoration is not accessible to disabled users.  Similarly, GNOME has no native support for third-party tray icons.  The only way to re-add such support is to write a shell extension, which is essentially monkey-patching the internals of GNOME Shell.

That said, GNOME should still be preferred if the UI/UX is sufficiently better *and* we can support it with our limited resources.  The first question is yours to answer @ninavizz, while the second is ultimately @marmarek’s as project lead.

**Comment by ninavizz** [comment]
@DemiMarie Of course—but I'd still love to hear @unman's thoughts to inform my ultimate recommendation. :)

**Comment by DemiMarie** [comment]
> @DemiMarie Of course—but I'd still love to hear @unman's thoughts to inform my ultimate recommendation. :)

Seems like KDE is better from a technical perspective, while GNOME is better from a UI/UX perspective.

**Comment by andrewdavidwong** [comment]
The "KDE vs. GNOME" discussion seems off-topic for this issue. I suggest taking it to a more appropriate place.

**Comment by ninavizz** [comment]
@unman Are you on the forums? I'd love to hear your thoughts on KDE as _an end user_ at some point, there, if you'd be keen! @DemiMarie we can pick this up in the forum.

**Comment by unman** [comment]
On Thu, Jun 17, 2021 at 11:14:28PM -0700, Nina Eleanor Alter wrote:
> @unman Are you on the forums? I'd love to hear your thoughts on KDE as _an end user_ at some point, there, if you'd be keen! @DemiMarie we can pick this up in the forum. 
> 

Yes - I'll post something there.
I think that KDE provides great UX for Qubes use - I'll try to explain
why there.

**Comment by marmarek** [comment]
I'd just share how I deal with large number of qubes and applications:
1. I don't use taskbar at all.
2. I use several workspaces (12 currently), bound to Super+F1-F12 keys (I think the default is Ctrl+Fx).
3. I have a system which application or qube belongs to which workspace, plus 2 workspaces reserved for things not fitting elsewhere (rarely used in practice, and only for short times). In most cases I have 2-3 windows per workspace, but there are some with up to 5.
4. Normally I switch between them with Super+Fx (mentioned above) and Alt+Tab
5. If I need an overview of all running applications, I search for an unobstructed desktop part and middle-click there - it gives list of all applications on all workspaces. There is probably some key combo for that, but I never managed to find/remember it.

**Comment by ninavizz** [comment]
WOW... Marek's brain. Just—WOW.

I'd still like there to be an "easier" method for users to manage multiple workspaces. 

Marek's brain is like the non-bionic equivalent to Steve Austin's body. 

I appreciate this issue existing, for the rest of us whom are mere mortals. And Marek's shared method, above, for a workaround until the funding falls from the sky to develop a better DE suited to Qubes.

**Comment by ddevz** [comment]
>     5. If I need an overview of all running applications, I search for an unobstructed desktop part and middle-click there - it gives list of all applications on all workspaces. There is probably some key combo for that, but I never managed to find/remember it.

I had no idea.  This makes a significant increase in the number before it becomes unmanageable (because it pulls down and does not have to fit in the taskbar).  Perhaps a hierarchical (like in her example) version of the middle click pull down menu would be a good answer.

**Comment by airelemental** [comment]
For window management using xfce, I do this:
1. about 5 workspaces, with VMs assigned to specific workspaces (using devilspie);
2. an icon-only taskbar (xfce window buttons widget with names turned off), showing only windows in the active workspace;
3. dropdown menu with a list of all windows across all workspaces (xfce Window Menu widget), for when I need to hunt down a specific window

**Comment by ddevz** [comment]
@marmarek 
>>     5. If I need an overview of all running applications, I search for an unobstructed desktop part and middle-click there - it gives list of all applications on all workspaces. There is probably some key combo for that, but I never managed to find/remember it.
>
>I had no idea. This makes a significant increase in the number before it becomes unmanageable (because it pulls down and does not have to fit in the taskbar). Perhaps a hierarchical (like in her example) version of the middle click pull down menu would be a good answer.

Update:
1. Go to qubes icon -> System tools -> panel -> Items -> click "add"
2. Select "window menu", "add", then "close"
3. Move the window menu up to be just below the "applications menu" (or wherever you want it in the panel)

Doing this makes it so you don't have to minimize windows to "search for an unobstructed desktop part".  Just click the new spot next to the qubes icon and the exact same menu he was referring to.  

This seems a really easy fix to increase the "18+ limit" some more, while waiting for a perfect UI to be decided on and implemented.

I'll let you guys decide if this should be set to default so that people other then just me can benefit from it.  It seems to me that for such a small amount of screen real estate, it would be a good deal for everyone with many windows, and a neutral deal for those who don't use many windows.
